indexing-service/src/main/java/io/druid/indexing/common/task/AppenderatorDriverRealtimeIndexTask.java

```diff
@@ -578,33 +441,158 @@ private void publishSegments(
-      public void onSuccess(@Nullable SegmentsAndMetadata published)
+      public void announceSegment(final DataSegment segment) throws IOException
       {
-        ListenableFuture<SegmentsAndMetadata> handoffFuture = driver.registerHandoff(published);
-
-        log.info("Registering pending handoff for [%s]", published);
+        // Side effect: Calling announceSegment causes a lock to be acquired
+        Preconditions.checkNotNull(
+            toolbox.getTaskActionClient().submit(
+                new LockAcquireAction(TaskLockType.EXCLUSIVE, segment.getInterval(), lockTimeoutMs)
+            ),
+            "Cannot acquire a lock for interval[%s]",
+            segment.getInterval()
+        );
+        toolbox.getSegmentAnnouncer().announceSegment(segment);
+      }
 
-        pendingHandoffs.add(handoffFuture);
+      @Override
+      public void unannounceSegment(final DataSegment segment) throws IOException
+      {
+        try {
+          toolbox.getSegmentAnnouncer().unannounceSegment(segment);
+        }
+        finally {
+          toolbox.getTaskActionClient().submit(new LockReleaseAction(segment.getInterval()));
+        }
+      }
 
-        publishResultFuture.set(published);
+      @Override
+      public void announceSegments(Iterable<DataSegment> segments) throws IOException
+      {
+        // Side effect: Calling announceSegments causes locks to be acquired
+        for (DataSegment segment : segments) {
+          Preconditions.checkNotNull(
+              toolbox.getTaskActionClient().submit(
+                  new LockAcquireAction(TaskLockType.EXCLUSIVE, segment.getInterval(), lockTimeoutMs)
+              ),
+              "Cannot acquire a lock for interval[%s]",
+              segment.getInterval()
+          );
+        }
+        toolbox.getSegmentAnnouncer().announceSegments(segments);
       }
 
       @Override
-      public void onFailure(@Nullable Throwable throwable)
+      public void unannounceSegments(Iterable<DataSegment> segments) throws IOException
       {
-        log.error(throwable, "Error occurred publishing segments");
-        publishResultFuture.setException(throwable);
+        try {
+          toolbox.getSegmentAnnouncer().unannounceSegments(segments);
+        }
+        finally {
+          for (DataSegment segment : segments) {
+            toolbox.getTaskActionClient().submit(new LockReleaseAction(segment.getInterval()));
+          }
+        }
       }
-    });
+    };
+  }
+
+  private void logUnusedConfiguration(RealtimeTuningConfig tuningConfig)
+  {
+    warnIfNotNull(tuningConfig.getWindowPeriod(), "windowPeriod is not used by this task");
+    warnIfNotNull(tuningConfig.getVersioningPolicy(), "versioningPolicy is not used by this task");
+    warnIfNotNull(tuningConfig.getRejectionPolicyFactory(), "rejectionPolicyFactory is not used by this task");
+    warnIfNotNull(tuningConfig.getPersistThreadPriority(), "persistThreadPriority is not used by this task");
```

Comment:
`tuningConfig.getPersistThreadPriority()` returns an integer which cannot be null. This is because this option has a default value.

Probably better to add a new type of tuningConfig which doesn't allow these configurations.